src/lib/services/edge-service.ts

```diff
@@ -10,30 +10,60 @@ export default class EdgeService {
 
     private apiTokenService: ApiTokenService;
 
+    private flagResolver: IFlagResolver;
+
     constructor(
         { apiTokenService }: { apiTokenService: ApiTokenService },
-        { getLogger }: Pick<IUnleashConfig, 'getLogger'>,
+        {
+            getLogger,
+            flagResolver,
+        }: Pick<IUnleashConfig, 'getLogger' | 'flagResolver'>,
     ) {
         this.logger = getLogger('lib/services/edge-service.ts');
         this.apiTokenService = apiTokenService;
+        this.flagResolver = flagResolver;
     }
 
     async getValidTokens(tokens: string[]): Promise<ValidatedEdgeTokensSchema> {
-        const activeTokens = await this.apiTokenService.getAllActiveTokens();
-        const edgeTokens = tokens.reduce((result: EdgeTokenSchema[], token) => {
-            const dbToken = activeTokens.find((activeToken) =>
-                constantTimeCompare(activeToken.secret, token),
-            );
-            if (dbToken) {
-                result.push({
-                    token: token,
-                    type: dbToken.type,
-                    projects: dbToken.projects,
-                });
+        if (this.flagResolver.isEnabled('checkEdgeValidTokensFromCache')) {
+            // new behavior: use cached tokens when possible
+            // use the db to fetch the missing ones
+            // cache stores both missing and active so we don't hammer the db
+            const validatedTokens: EdgeTokenSchema[] = [];
+            for (const token of tokens) {
+                const found =
+                    await this.apiTokenService.getTokenWithCache(token);
+                if (found) {
+                    validatedTokens.push({
+                        token: token,
+                        type: found.type,
+                        projects: found.projects,
+                    });
+                }
```

Comment:
New behavior checking the tokens one by one, hitting the cache unless one is not present. Invalid tokens are also cached